mobilecoind-dev-faucet/src/slam/mod.rs

```diff
@@ -0,0 +1,500 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+use super::{GetUtxoError, UtxoRecord, Worker};
+use displaydoc::Display;
+use mc_account_keys::{AccountKey, PublicAddress};
+use mc_common::logger::{log, o, Logger};
+use mc_mobilecoind_api::mobilecoind_api_grpc::MobilecoindApiClient;
+use mc_util_uri::ConsensusClientUri;
+use std::{
+    sync::{
+        atomic::{AtomicBool, AtomicU64, AtomicUsize, Ordering},
+        Arc,
+    },
+    time::{Duration, Instant},
+};
+
+mod prepared_utxo;
+use prepared_utxo::PreparedUtxo;
+
+mod tx_submitter;
+use tx_submitter::{SubmitTxError, TxSubmitter};
+
+/// Parameters to start a slam operation
+#[derive(Clone, Debug)]
+pub struct SlamParams {
+    /// The total number of tx to try to submit
+    pub target_num_tx: usize,
+    /// The number of threads to use when submitting to consensus
+    pub num_threads: usize,
+    /// The number of retries to allow when submitting to consensus
+    pub retries: usize,
+    /// The period to wait between retries when submitting to consensus
+    pub retry_period: Duration,
+    /// How far forward to set the tombstone block
+    pub tombstone_offset: u64,
+    /// The consensus uris to submit to
+    pub consensus_client_uris: Vec<ConsensusClientUri>,
+}
+
+impl Default for SlamParams {
+    fn default() -> Self {
+        Self {
+            target_num_tx: 500,
+            num_threads: 30,
+            retries: 30,
+            retry_period: Duration::from_millis(1000),
+            tombstone_offset: 10,
+            consensus_client_uris: Default::default(),
+        }
+    }
+}
+
+/// A report on a completed slam operation
+#[derive(Clone, Debug, Default)]
+pub struct SlamReport {
+    /// The number of utxos successfully prepared
+    pub num_prepared_utxos: usize,
+    /// The number of txs successfully submitted to the network
+    pub num_submitted_txs: usize,
+    /// The total time spent preparing Txs
+    pub prepare_time: Duration,
+    /// The total time spent submitting Txs
+    pub submit_time: Duration,
+}
+
+/// State which tracks the progress of a running slam operation.
+/// This allows one thread to conduct the slam while another thread checks on
+/// its status.
+pub struct SlamState {
+    // Phase is a state variable:
+    // 0: Not currently slamming
+    // 1: Connecting
+    // 2: Preparing utxos: Obtaining proofs of membership
+    // 3: Submitting transactions
```

Comment:
Might be easier to follow if these were made into consts (e.g. `STATE_INACTIVE`, `STATE_CONNECTING`, etc)